summoner-cli/src/Summoner/Project.hs

```diff
@@ -155,25 +157,61 @@ generateProject settingsNoUpload isOffline projectName Config{..} = do
         Last prelude@(Just (CustomPrelude p _)) ->
             prelude <$ successMessage ("Custom prelude " <> p <> " will be used in the project")
 
+    getNixPkgSet :: Bool -> IO (Maybe NixPkgSet)
+    getNixPkgSet usingNix = if usingNix
+        then case cNixPkgSet of
+            Last Nothing -> do
+                let yesDo, noDo :: IO (Maybe NixPkgSet)
+                    yesDo = do
+                        o <- queryNotNull "Nix package set GitHub owner name: "
```

Comment:
*Sorry, I will ask a lot of (stupid) questions, as I'm not familiar with nix.*

Is that could be any name or just limited amount of possible answers here? I noticed that in your example config you have `nixOS`, should this be the default?

Reply:
Don't worry about questions you ask. There are two ways you can use nix w.r.t. dependencies; using an unpinned (system-level) package set, that comes with an installation of nix/nixos, or a pinned package set. the advantage to a pinned package set is that builds become deterministic across multiple machines, since there can be no guarantee that the package set i use is the same as the one you use. The way a nix package set is usually acquired is through github. There are four fields to a nix package set:

* Owner (GitHub owner name)
* Repo (GitHub repo name)
* Revision (GitHub SHA1 revision)
* SHA256 Hash of the package set

for owner $OWNER, repo $REPO, revision $REVISION, and SHA256 hash $SHA, the url for the package set would be at https://github.com/$OWNER/$REPO/archive/$REVISION.tar.gz

You may be wondering where the SHA comes into play. When you attempt to use a package set, you must provide a SHA256 hash of the package set's unpacked NAR archive (more on NAR in a bit). Nix checks that the SHA256 you provide is equivalent to the SHA256 of the package set, and if it is not equivalent, rejects the package set and fails to build. This ensures that only trusted package sets are used. SHA1 is not suitable for this, as has been proven multiple times by the NPM/Golang community. I think a recent feature of stack is to also allow for the sha256 hash of github dependencies, and iirc cabal will soon follow suit, if it hasn't already.

A NAR is a Nix ARchive file. NAR is a (i think strict) subset of tar. TAR, and other archivers, are not deterministic. They add variable-length padding, do not sort files, add timestamps, and a bunch of other stuff. Thus a SHA256 hash of a TAR archive is insufficient. See section 5.2 of eelco's [thesis](https://nixos.org/~eelco/pubs/phd-thesis.pdf).

For the purposes of summoner, I provide a default as `Summoner.Settings.defaultNixPkgSet`, which uses a specific git revision and hash.

You may be wondering how users compute the hash? I touched on this rather vaguely in a comment yesterday. There are two ways.

1) Insert a fake hash with the correct length and nix will complain that it received the wrong hash, and halt the build. It will provide the correct hash. This method is slow and annoying, since it requires you to trigger a build. I don't recommend it normally.
2) use `nix-prefetch-url`, an executable that can give you a bunch of information about urls w.r.t. how they can be used by nix. the most common use case (and the one we might care about) is:

```sh
nix-prefetch-url --unpack https://github.com/$OWNER/$REPO/archive/$REVISION.tar.gz
```

This will take the given nix package set found at that github archive and return its SHA256 hash. The --unpack flag tells it to give you the hash of the unpacked NAR archive. To get the hash of the TAR archive, remove that flag. The comment I made yesterday was saying that to make it easier on users, a potential enhancement is that they only need to provide a revision, and we will compute the hash for them. I have a simple haskell tool that already does this [here](https://github.com/chessai/grab). This feature is not necessary or integral, just an enhancement.